avalon/pipeline.py

```diff
@@ -926,61 +926,103 @@ def get_representation_context(representation):
     return context
 
 
-def update_current_task(task=None, asset=None, app=None):
-    """Update active Session to a new task work area.
+def compute_session_changes(session, task=None, asset=None, app=None):
+    """Compute the changes for a Session object on asset, task or app switch
 
-    This updates the live Session to a different `asset`, `task` or `app`.
+    This does *NOT* update the Session object, but returns the changes
+    required for a valid update of the Session.
 
     Args:
-        task (str): The task to set.
-        asset (str): The asset to set.
-        app (str): The app to set.
+        session (dict): The initial session to compute changes to.
+            This is required for computing the full Work Directory, as that
+            also depends on the values that haven't changed.
+        task (str, Optional): Name of task to switch to.
+        asset (str or dict, Optional): Name of asset to switch to.
+            You can also directly provide the Asset dictionary as returned
+            from the database to avoid an additional query. (optimization)
+        app (str, Optional): Name of app to switch to.
 
     Returns:
-        dict: The changed key, values in the current Session.
+        dict: The required changes in the Session dictionary.
 
     """
 
-    mapping = {
-        "AVALON_ASSET": asset,
-        "AVALON_TASK": task,
-        "AVALON_APP": app,
-    }
-    changed = {key: value for key, value in mapping.items() if value}
-    if not changed:
-        return
+    changes = dict()
 
-    # Update silo when asset changed
-    if "AVALON_ASSET" in changed:
-        asset_document = io.find_one({"name": changed["AVALON_ASSET"],
-                                      "type": "asset"})
-        assert asset_document, "Asset must exist"
-        changed["AVALON_SILO"] = asset_document["silo"]
+    # If no changes, return directly
+    if not any([task, asset, app]):
+        return changes
+
+    if task:
+        changes["AVALON_TASK"] = task
+
+    if app:
+        changes["AVALON_APP"] = app
+
+    # Update silo and hierarchy when asset changed
+    if asset:
+        if isinstance(asset, dict):
```

Comment:
This has now additionally been allowed to be passed a database document for an asset as opposed to only working by its name. This allows to avoid the additional query to the database.